### ISSUE TITLE
Balance Suggestion: Brick crafting following charcoal crafting mechanic

### ISSUE THREAD

**Issue by glenmack** [CleverRaven/Cataclysm-DDA#20420]
Brick crafting currently takes 40 minutes, and produces 1/30th of the bricks needed for 1 tile of terrain.

The balance as far as resources goes seems fine, but the personal time investment (not the time needed, important note) is unrealistic, a grind and unfun.

It uses the brick Kiln, which must be reloaded in your inventory with charcoal. This immediately overburdens most characters as a fully loaded kiln weighs 230kg. It feels like a very odd step in the game process.

Almost no method of hand crafting bricks involves watching the firing process. (Even during the pre-industrial era, you might pay a young boy to watch it didn't burn the whole place down but not watch the process itself.) It usually occurs over a much longer time, with much larger quantities of bricks.

http://www.openairclassroom.org.uk/Further%20information/information-making%20bricks%20by%20hand.htm

My suggestion is that the code for making charcoal (placing logs in a kiln based piece of furniture and igniting) be adapted for bricks. If it is easy to load 40 clay and 2400 charcoal in to a single piece of furniture then fire and leave for 24 hours, great. Remove the brick recipe and it would provide better balance and a better user experience.

I'm not a coder, but I'm guessing it probably works on the principle of guns and other ammo loading in the game 1 thing in = proportional amount out. Which would make loading with charcoal and unfired bricks difficult to code.

If this is the case I suggest a quick recipe change, brick crafting would become unfired brick and take ~ 1-2 minutes, or produce 40 unfired bricks in 40 minutes, then add a recipe for brick pallet; 40 bricks, 4 2x4 planks and 2400 charcoal. Loading that into the kiln would then produce the appropriate amount of bricks.

In summary: Player time crafting is far too long and boring for bricks. While the time seems proportionally appropriate, the idea that the player should be actively involved is unrealistic and grind-y.

**Comment by Coolthulhu** [comment]
I see quite a bit of "use kiln style crafting for x crafting" suggestions recently.

I oppose the idea of copypasting that kiln code for new things, but if someone were to write a jsonable "converter" furniture which would cover kilns, fermenting vats and the new mechanics (smoking/drying meat, baking bricks, composting organic waste etc.), this could get in.

**Comment by glenmack** [comment]
If someone were to go down the jsonable "converter" route, as suggested, which seems like a fairly good way to go, I would also suggest a skill related component checked at time of loading the furniture. At least as far a bricks go a "failure rate" is always present at any skill level and would probably apply to many other "unsupervised batch processes".

**Comment by DangerNoodle** [comment]
If such a thing is reworked, it would be nice if this allows for the furniture to be loaded with more than just the default ammunition for the ammo type it uses. Reworking the general behavior of charcoal-fueled furniture would be a good opportunity to implement that as well.

**Comment by Veraghin** [comment]
"If such a thing is reworked, it would be nice if this allows for the furniture to be loaded with more than just the default ammunition for the ammo type it uses. Reworking the general behavior of charcoal-fueled furniture would be a good opportunity to implement that as well."

What do you mean by the furniture to be loaded?

Also, I've started looking at the code. Changing the "examine_action" to the new "converter" and handling everything in there should work. For the jsonization, I'm wondering whether to stick the new information into furniture.json, or start another json file, with lookup by furniture id, containing information such as fuels, reagents, skills used, resulting items etc. for each "converter" furniture.

**Comment by glenmack** [comment]
> What do you mean by the furniture to be loaded?

I think he means that instead of having a charcoal kiln, brick kiln, clay kiln etc. all doing one specific conversion each, you have one "kiln" that does conversions based on what is loaded into it.

**Comment by Coolthulhu** [comment]
> For the jsonization, I'm wondering whether to stick the new information into furniture.json, or start another json file, with lookup by furniture id, containing information such as fuels, reagents, skills used, resulting items etc. for each "converter" furniture.

Would be better as another file. Something like a recipe, but simpler. Referenced by ID.
Otherwise each converter json would get bloated fast.

**Comment by DangerNoodle** [comment]
> What do you mean by the furniture to be loaded?

I mean the issue of charcoal forges only being able to use charcoal, not coal: https://github.com/CleverRaven/Cataclysm-DDA/issues/18329

**Comment by Veraghin** [comment]
Okay, the way I think it could work:
Each furniture that can act as a "converter" will call an iexamine::converter function, which will take input from a processes.json file. Converters would be those pieces of furniture which can be left alone to do their thing, such as the kiln, the fermenting vat, the smoking rack,maybe the crucible etc.
There will be a new json file, processes.json, where we'll have recipe style blocks detailing the processes, referenced by the furniture id. 
It would contain:
- the furniture id
- the furniture we transform this one into,(kiln_empty -> kiln_full)
- possible input and intermediate step pairs (wood/bone -> unfinished charcoal, chunk of meat would stay as it is, as would chunks of steel etc)
- the possible fuel types
- how much fuel do we need per reagent, if any
- the messages we get in case the furniture is empty, full, has some type of reagent in it and whether we want to start the process or not, and a message confirming starting of the process
- the maximum volume of reagents we can put in the furniture

Multi-step processes could be implemented as the furniture transforming into intermediate steps, calling different recipes before a final processing.

Also, the furniture we transformed into would handle the 
- process duration
- output item based on input
- messages detailing process finish etc

Any thoughts? I'll start on this I get the knack of handling json files.

**Comment by Coolthulhu** [comment]
Overall good, but I have some suggestions regarding the structure:

> chunk of meat would stay as it is

Minor: for chunks of meat to stay properly usable, either all the results would need to be guaranteed permafood or something would need to be done about rot.

> how much fuel do we need per reagent, if any

Would be better to have a more recipe-like format: for example 1 meat and 3 charcoal produces 1 smoked meat. Then it would be more familiar and consistent.

> the messages we get in case the furniture is empty, full, has some type of reagent in it and whether we want to start the process or not, and a message confirming starting of the process

This is getting dangerously similar to iuse_actor mechanics, except with data shifted off to a different structure. Not an error, but a warning that the wheel is being reinvented.
Would be better if those messages were next to "converter" designation than the recipes themselves, if possible, so that one converter recipe could belong to more than one converter.
That would require some more major restructuring of `iexamine`, so at first the fully separated converter structure could be better, I guess.

> the maximum volume of reagents we can put in the furniture

This is already handled by limiting the volume at furniture level.

> Also, the furniture we transformed into would handle the

Would be better to handle them all in the one process structure, with actual furniture type only controlling the current state.

**Comment by Veraghin** [comment]
> Minor: for chunks of meat to stay properly usable, either all the results would need to be guaranteed permafood or something would need to be done about rot.

I guess I was only thinking of the smoke rack, where the output is permafood, obviously this could be changed process by process.

> Would be better to have a more recipe-like format: for example 1 meat and 3 charcoal produces 1 smoked meat. Then it would be more familiar and consistent.

Will do.

> This is getting dangerously similar to iuse_actor mechanics

I don't like it either, but this is the most straightforward solution.

> Would be better to handle them all in the one process structure, with actual furniture type only controlling the current state.

I'll see what I can come up with. 

Also, could you point me in the right direction regarding json files and Json Objects? I can't seem to find the header file for the  .obj() function called for example in "auto furniture = g->m.furn( location ).obj() , and I guess I'd have to use an analogue to that to load this new json file.

**Comment by Coolthulhu** [comment]
> Also, could you point me in the right direction regarding json files and Json Objects? I can't seem to find the header file for the .obj() function called for example in "auto furniture = g->m.furn( location ).obj() , and I guess I'd have to use an analogue to that to load this new json file.

That's string ids. They can be pretty complex at first, but thankfully there is a great example of something similar to what you'd be doing: `gates.h/cpp` - it has pretty much all what you'd need to set up regarding the ids, generic factory and loading.

**Comment by Veraghin** [comment]
I'm making headway.
The way it's shaping up now, we'll have "converters"/processors all leading to the same iexamine action, which will handle the process part.  Should it work by checking items dropped on the processor tile,  or check the player's inventory for items? Should this depend on the process?

**Comment by Coolthulhu** [comment]
I'd say dropped.
If it was in player's inventory, you'd need to invent a good menu to ask for components.
A benefit of asking the player would be that you could get only what you want. But it would be more painful for player. And you still don't avoid sanitization: you need to detect if you have only what you want at the end anyway.

I see the algorithm/use case working as follows:
* Player drops crap on converter
* Player examines converter
* Converter goes through all processes until it finds one that can be performed with current ingredients
* If it can't find any, it describes all the possible processes of this converter somehow (at first it would be good enough if you just hardcoded/jsoned a string for that - describing stuff programatically is surprisingly hard)
* If it has found one, it says which one will it use. Bonus points for saying how much will be produced
* Player says yes/no
* If player says yes, we go into processing. All ingredients need to have their birthdays adjusted or otherwise there has to be some mark that the process started in the current turn.
* Once processing is done and player examines the converter, go through all the processes, pick the first one and convert as much as you can
* Exit the converter examine function
* Player manually removes the stuff from converter (or not, if the converter can handle that)

This gets simpler if only one process is possible.

It's a good idea to PR when you're ~75% done, with a [WiP] tag. That way we'll be able to help with the structure before the last moment and before you fully commit to some specific solution.

**Comment by Veraghin** [comment]
> If player says yes, we go into processing. All ingredients need to have their birthdays adjusted or otherwise there has to be some mark that the process started in the current turn.

If I modify the birthday of something that can rot, can that cause it to become unrotten?

**Comment by kevingranade** [comment]
Please convert just one thing to use the new system, ideally the simplest
thing that needs it, and post that as a PR before converting everything
else.

**Comment by DangerNoodle** [comment]
The simplest as least intrusive first step there might be brick kilns. A charcoal smoker would involve a lot more recipes, dealing with perishable ingredients, the fact that many recipes allow using a charcoal smoker as an alternative to a fire (conventional cooking might not warrant this yet), and other possible complications.

Brick kilns might be the easiest way to test this, with the fewest complications and the least potential disruption to player habits.

**Comment by glenmack** [comment]
No progress after a year? If this idea has been shuttered, please inform me.

**Comment by illi-kun** [comment]
Nobody works on code for this, so, yeah, no progress. You can try to implement it by yourself. Elsewise, all you can is just wait for someone to do this (sooner or later).

**Comment by ZhilkinSerg** [comment]
> No progress after a year? If this idea has been shuttered, please inform me.

If you want something done - start doing it by yourself. There are [174 open suggestion issues at the moment](https://github.com/CleverRaven/Cataclysm-DDA/labels/%3CSuggestion%20%2F%20Discussion%3E) and 500 something other issues and we have very limited developer resources.
